src/coreclr/src/jit/importer.cpp

```diff
@@ -4015,6 +4015,75 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Type_get_IsClass:
+            case NI_System_Type_get_IsPrimitive:
+            case NI_System_Type_get_IsValueType:
+            {
+                // Optimize things like
+                //
+                //   ldtoken [Type]
+                //   call Type.GetTypeFromHandle (which is replaced with CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE)
+                //   call Type.IsValueType (also IsClass and IsPrimitive)
+                //
+                // to `true` or `false`
+                // e.g. `typeof(int).IsValueType` => `true`
+                if (impStackTop().val->IsCall())
+                {
+                    GenTreeCall* call = impStackTop().val->AsCall();
+                    if (call->gtCallMethHnd == eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE))
+                    {
+                        GenTree* arg = call->Args().begin()->GetNode();
+                        if (!arg->IsIntegralConst())
```

Comment:
This should use `gtGetHelperArgClassHandle` or something similar so that it kicks in for handles with relocations, etc.